### PR TITLE
Menubar - Add "find menu item" search feature

### PR DESCRIPTION
**Pull request by colemanw**
Overview
----------------------------------------
Allows user to locate menu items by typing a few letters.
Find tool is located under the "Home" (Civi logo) menu.

Before
----------------------------------------
No finder for menu items.

After
----------------------------------------
The first item in the "Home" menu is a finder to locate menu items. Accesskey "M"

Screenshot from 2020-02-21 18-45-46

Typing a few letters reveals a sub-menu of clickable results:

Screenshot from 2020-02-21 18-46-35

Comments
----------------------------------------
This obsoletes the [QuickSearch Extension](https://civicrm.org/extensions/quick-search-civicrm-menu).

### CODE REVIEW

**Comment by twomice** [comment]
So. Freakin. Cool.

**Comment by eileenmcnaughton** [comment]
@twomice any chance you can give it a spin? Code passes the 'nothing looks worrying test'

**Comment by twomice** [comment]
Sure, will give it a go.

**Comment by twomice** [comment]
@colemanw If you can ping me when tests are passing, I'll give it a review.

**Comment by colemanw** [comment]
@twomice the 4 test failures are unrelated. Those look like flaky time-based tests which fail at certain times of the day/week/month.

**Comment by twomice** [comment]
@colemanw When I type, an "x" control appears in the text field, but that x does nothing when I click on it. (Otherwise looks good so far.)

**Comment by colemanw** [comment]
@twomice I just noticed that too. Looks to be a side-effect of the menu's markup and the `<input>` being nested within an `<a>` tag which is probably not valid html. It also caused this other problem: https://lab.civicrm.org/dev/core/issues/1525

The "right" fix is somewhat elusive but for now I've just changed the input type from "search" to "text" so the non-functional `x` won't appear.

**Comment by eileenmcnaughton** [comment]
test this please

**Comment by MegaphoneJon** [comment]
I know that full keyboard shortcut support is a heavy lift for Civi at the moment - but based on my use of the [Coffee](https://www.drupal.org/project/coffee) module for Drupal, I'm guessing that if this had a keyboard shortcut it would quickly become a favorite tool of Civi power users everywhere.

**Comment by colemanw** [comment]
@MegaphoneJon great idea! I looked into it and actually there is already a shortcut key to the "Home" (Civi logo) menu (ALT + M) which wasn't really doing anything except opening that item. I just moved the search box to that menu item (arguably a better place for it) and now the ALT + M hotkey will take you straight to the finder. (note the hotkey could be ALT+SHIFT or something else depending on the browser/OS).

**Comment by twomice** [comment]
One problem with using ALT+M for this is that now I can't access "hide civicrm menu" via the keyboard, nor or other sub-items of the "Civi logo" menu. Not a deal breaker as far as I'm concerned, but it does have the effect of removing all o the previous utility of this shortcut in favor of different functionality.

I don't know how we could measure the potential impact of this change (we probably don't have usage stats on  this shortcut key), but thought it worth mentioning. @colemanw , any thoughts?

**Comment by colemanw** [comment]
@twomice I just tried it and was able to access the "hide menu" item with `ALT + M` `TAB` `TAB`.

**Comment by twomice** [comment]
I'm cool with that if you are. Reviewing now.

File: extension-compatibility.json
Comment (twomice) [inline extension-compatibility.json]: @colemanw Wasn't aware of "obsolete" capability. Kinda cool. Is this documented somewhere?

**Review by twomice — APPROVED** [review]
(*CiviCRM Review Template WORD-1.2*)



* General standards
    * ([`r-explain`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-explain)) Pass: Explanation is clear and thorough.
    * ([`r-user`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-user)) Pass: Minor usability impact (lost use of arrow keys to access sub-menu items with ALT+M shortcut key) is mitigated with workaround (use of Tab/Shift+Tab instead)
    * ([`r-doc`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-doc)) Pass: ALT+M Shortcut key is documented in "Access Keys" help popup.
    * ([`r-run`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-run)) Pass: I used this feature on the auto-generated test website (http://core-16597-6w4xq.test-1.civicrm.org:8001/ at time of writing) and observed it works as expected.
* Developer standards
    * ([`r-tech`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-tech)) Pass: Additional UI-only functionality with dfeined and limited scope, not impacting existing features.
    * ([`r-code`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-code)) Pass
    * ([`r-maint`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-maint)) Pass
    * ([`r-test`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-test)) Pass

**Comment by twomice** [comment]
@eileenmcnaughton See my review above. I believe this is ready to merge.

File: extension-compatibility.json
Comment (colemanw) [reply extension-compatibility.json]: No it's fairly new. So far its being used for extensions whose features have been migrated into core (Api4, KAM, ExportUI, etc.). Adding an extension to this list will:

1. Automatically disable/uninstall it during the upgrade.
2. Not allow it to be reenabled; label it "obsolete" in the UI.
3. Resolve any dependencies (so an extension that declares a dependency on Api4 will still work)

Let me see if I can find a spot for that in the dev docs pages...

Comment (colemanw) [reply extension-compatibility.json]: https://github.com/civicrm/civicrm-dev-docs/pull/764

**Comment by eileenmcnaughton** [comment]
I've given this merge-ready - I just want to give @aydun & @michaelmcandrew - who was the other person who was involved in usability over your way? a chance to see the notes about the keyboard shortcut changes

**Comment by michaelmcandrew** [comment]
hey @eileenmcnaughton - it was claire williams from www.visibility.org.uk - I've sent her a quick email.

**Comment by aydun** [comment]
I like the new feature - slight concern re the arrow keys as documented above but doesn't seem a blocker to me.

Occasionally it brings up duplicates: eg entering 'cam' brings up 4 links for Personal Campaign Pages, 2 each for events and contribute - but I'm guessing that is related to the underlying menu items under Contributions and Events being duplicated under Administer.

Also looping in @JoeMurray re accessibility.

**Comment by colemanw** [comment]
I get the concern about the arrow keys but I'm not sure anything can be done about it. The issue is that the search input is a text field, and pretty much every UI convention says that when a text field is focused, the arrow keys are for moving the cursor, not for moving focus to a different element. FYI the quicksearch field has the exact same issue.

And yes, the menubar is full of duplicates. Probably to make things easier to find... because of the lack of a search box!

**Comment by seamuslee001** [comment]
This looks good to me as well and will be a nice improvement

**Comment by yashodha** [comment]
@colemanw @seamuslee001 tested this and works well :+1: 
report

**Comment by yashodha** [comment]
@colemanw Shall we merge or we waiting on more feedback/changes?

**Comment by colemanw** [comment]
I think this is ready to merge.

**Comment by eileenmcnaughton** [comment]
OK - I'll merge - if we get further feedback from @michaelmcandrew's email  / pings we can address

**Pull request merged**
